metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -191,13 +212,75 @@ ::call_outgoing_kmers(node_index node, const OutgoingEdgeCallback &callback) con
     }
 }
 
+void CanonicalDBG
+::append_parent_nodes_using_node_rev_comp(node_index node,
+                                          std::vector<node_index> &parents) const {
+    /**
+     * find parents of node by searching for children of its reverse complement
+     * e.g., node = AGCCAT. Find TAGCCA and AAGCCA by looking for TGGCTA and TGGCTT.
+     *  TAGCCA                    TGGCTA
+     *        \                  /
+     *         AGCCAT  ->  ATGGCT
+     *        /                  \
+     *  AAGCCA                    TGGCTT
+     */
+
+    const auto &alphabet = graph_.alphabet();
+
+    //        lshift    rc
+    // AGCCAT -> *AGCCA -> TGGCT*
+    std::string rev_seq = std::string(1, '\0') + get_node_sequence(node).substr(0, get_k() - 1);
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    assert(rev_seq.back() == '\0');
+
+    // for each n, check for TGGCTn. If found, define and store the index for
+    // rc(n)AGCCA as index(TGGCTn) + offset_
+    const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(&graph_);
+    if (dbg_succ) {
+        // Find the BOSS node TGGCT and iterate through all of its outdoing edges.
+        // Then, convert the edge indices to get the DBGSuccinct node indices
+        const auto &boss = dbg_succ->get_boss();
+        auto encoded = boss.encode(std::string_view(rev_seq.data(), get_k() - 1));
+
+        auto [edge, edge_2, end] = boss.index_range(encoded.begin(), encoded.end());
+
+        if (edge && edge_2 && end == encoded.end()) {
+            assert(edge == edge_2);
+            boss.call_outgoing(edge, [&](auto adjacent_edge) {
+                node_index prev = dbg_succ->boss_to_kmer_index(adjacent_edge);
+                if (prev) {
+                    char c = alphabet[boss.get_W(adjacent_edge) % boss.alph_size];
+                    if (c == boss::BOSS::kSentinel)
+                        return;
+
+                    size_t i = boss.encode(::reverse_complement(c));
+
+                    if (parents[i] == DeBruijnGraph::npos)
+                        parents[i] = prev + offset_;
+                }
+            });
+        }
+
+    } else {
+        for (size_t i = 0; i < alphabet.size(); ++i) {
+            if (parents[i] != DeBruijnGraph::npos)
+                continue;
+
+            rev_seq.back() = ::reverse_complement(alphabet[i]);
+            node_index prev = graph_.kmer_to_node(rev_seq);
+            if (prev != DeBruijnGraph::npos)
+                parents[i] = prev + offset_;
+        }
+    }
+}
+
 void CanonicalDBG
 ::call_incoming_kmers(node_index node, const IncomingEdgeCallback &callback) const {
     assert(node);
     assert(node <= offset_ * 2);
     if (node > offset_) {
         call_outgoing_kmers(node - offset_, [&](node_index prev, char c) {
-            ::reverse_complement(&c, &c + 1);
+            c = ::reverse_complement(c);
```

Comment:
Maybe we should call it `complement` instead of `reverse_complement`, and then you wouldn't have to add `::` in front.
```suggestion
            c = complement(c);
```